clients/spark/core/src/main/scala/io/treeverse/clients/SSTableReader.scala

```diff
@@ -66,29 +73,45 @@ object SSTableReader {
     localFile
   }
 
-  def forMetaRange(configuration: Configuration, metaRangeURL: String): SSTableReader[RangeData] = {
+  def forMetaRange(
+      configuration: Configuration,
+      metaRangeURL: String,
+      dontOwn: Boolean = false
+  ): SSTableReader[RangeData] = {
     val localFile: File = copyToLocal(configuration, metaRangeURL)
-    new SSTableReader(localFile, RangeData.messageCompanion)
+    val ret = new SSTableReader(localFile, RangeData.messageCompanion, dontOwn)
+    Option(TaskContext.get()).foreach(_.addTaskCompletionListener[Unit](_ => ret.close()))
+    ret
   }
 
-  def forRange(configuration: Configuration, rangeURL: String): SSTableReader[Entry] = {
+  def forRange(
+      configuration: Configuration,
+      rangeURL: String,
+      dontOwn: Boolean = false
+  ): SSTableReader[Entry] = {
     val localFile: File = copyToLocal(configuration, rangeURL)
-    new SSTableReader(localFile, Entry.messageCompanion)
+    val ret = new SSTableReader(localFile, Entry.messageCompanion, dontOwn)
+    Option(TaskContext.get()).foreach(_.addTaskCompletionListener[Unit](_ => ret.close()))
+    ret
   }
 }
 
 class SSTableReader[Proto <: GeneratedMessage with scalapb.Message[Proto]] private (
     val file: java.io.File,
-    val companion: GeneratedMessageCompanion[Proto]
+    val companion: GeneratedMessageCompanion[Proto],
+    val dontOwn: Boolean = false
 ) extends Closeable {
   private val fp = new java.io.RandomAccessFile(file, "r")
   private val reader = new BlockReadableFile(fp)
 
-  def this(sstableFilename: String, companion: GeneratedMessageCompanion[Proto]) =
-    this(new java.io.File(sstableFilename), companion)
+  def this(sstableFilename: String, companion: GeneratedMessageCompanion[Proto], dontOwn: Boolean) =
+    this(new java.io.File(sstableFilename), companion, dontOwn)
 
   def close(): Unit = {
     fp.close()
+    if (!dontOwn) {
```

Comment:
why so negative?
can we change it to `own` in all places?

Reply:
Fine, be positive like that.

!["Think like a proton, always positive" motivational image](https://i.pinimg.com/originals/f4/58/a9/f458a9d1c6fe31a6e34fdcabc4866480.png)

Done.